consensus/src/consensus_observer/observer.rs

```diff
@@ -881,6 +908,66 @@ impl ConsensusObserver {
             })
     }
 
+    /// Verifies the ordered blocks and returns an error if the data is invalid.
+    /// Note: this does not check the ordered proof.
+    fn verify_ordered_blocks(
+        &self,
+        blocks: &[Arc<PipelinedBlock>],
+        ordered_proof: &LedgerInfoWithSignatures,
+    ) -> Result<(), Error> {
+        // Verify that we have at least one ordered block
+        if blocks.is_empty() {
+            return Err(Error::InvalidMessageError(
+                "Received empty ordered block!".to_string(),
+            ));
+        }
+
+        // Verify the last block ID matches the ordered proof block ID
+        if blocks.last().unwrap().id() != ordered_proof.commit_info().id() {
+            return Err(Error::InvalidMessageError(
+                format!(
+                    "Last ordered block ID does not match the ordered proof ID! Number of blocks: {:?}, Last ordered block ID: {:?}, Ordered proof ID: {:?}",
+                    blocks.len(),
+                    blocks.last().unwrap().id(),
+                    ordered_proof.commit_info().id()
+                )
+            ));
+        }
+
+        // Verify the blocks are correctly chained together (from the last block to the first)
+        let mut expected_parent_id = None;
+        for block in blocks.iter().rev() {
+            if let Some(expected_parent_id) = expected_parent_id {
+                if block.id() != expected_parent_id {
+                    return Err(Error::InvalidMessageError(
+                        format!(
+                            "Block parent ID does not match the expected parent ID! Block ID: {:?}, Expected parent ID: {:?}",
+                            block.id(),
+                            expected_parent_id
+                        )
+                    ));
+                }
+            }
+
+            expected_parent_id = Some(block.parent_id());
+        }
+
+        Ok(())
+    }
+
+    /// Verifies the ordered proof and returns an error if the proof is invalid
+    fn verify_ordered_proof(&self, ordered_proof: &LedgerInfoWithSignatures) -> Result<(), Error> {
```

Comment:
this can also be a verify function with an arg of &ValidatorVerifier

Reply:
SGTM. Will update 😄